Update README.md and Scenarios page (#641)
liqotech/liqo: docs/pages/User/Install/pre-install.md

---
title: Pre-Install
weight: 2
---

### Introduction

Liqo can be installed either in private or local clusters. Its configuration depends on the type of connectivity between the two clusters. Before installing Liqo, you have to consider how your clusters can connect to each other and can peer together.

### Peering Requirements

Liqo requires the following services to be reciprocally accessible to perform cluster peerings:

> * **Authentication server** used to authenticate other clusters (i.e. `liqo-auth`).

Reviewer: ```suggestion
* **Authentication server**: Liqo service, used to authenticate incoming peering requests coming from other clusters (i.e. `liqo-auth`).
```

> * The **Kubernetes API server** you want to peer.

Reviewer: ```suggestion
* **Kubernetes API server**: Kubernetes standard service, which has to be customized with the creation of some new resources when the peering process starts.
```

> * **Network gateway** used to establish interconnection between clusters (i.e. `liqo-gateway`)

Reviewer: ```suggestion
* **Network gateway**: Liqo service that take cares of setting up the inter-cluster connectivity between clusters (i.e. `liqo-gateway`)
```

> Those services have to be accessible from the other clusters to peer with them. This may change the way you would like to have them exposed.

Reviewer: ```suggestion
Depending on the physical setup of your cluster, you need to properly configure some parameters required by Liqo during the install process in order to enable remote clusters to contact the above services. 
```

Below it is possible to find some common scenarios that Liqo can handle. Once you identify yours, you can go have to the *table* of each section to find the right values you should specify when installing Liqo using the chart.

> 

Reviewer: I would add a link to the page explaining the Helm chart, or at least mentioning here which variables have to be updated in that chart.

> ### Cloud to cloud

Reviewer: I do not fully understand this section. Are those parameters set during the *setup* of the cluster, or during the *creation* of the peering? In case of the latter, I would propose to make this more clear in the first part of this document.

### Peering Requirements

Liqo requires the following services to be reciprocally accessible on both clusters to be able to start the cluster peering:

* **Authentication server**: Liqo service, used to authenticate incoming peering requests coming from other clusters (i.e., `liqo-auth`). You should modify the values in section ``auth`` in the [Liqo chart values](/user/install/chart_values) to configure how the authentication server is exposed.
* **Kubernetes API server**: Kubernetes APIServer, where the (remote) Liqo instance will create some new resources when the peering process starts. APIServer can be configured in the ``apiServer`` section of the [Liqo chart values](/user/install/chart_values). By default, Liqo will use an endpoint composed of the IP of the first master and the 6443 port. In managed clusters, you have to configure those values to have Liqo working correctly.
* **Network gateway**: Liqo service responsible for setting up the inter-cluster connectivity between clusters (i.e., `liqo-gateway`). The Liqo Gateway is configured in the ``gateway`` section of the [Liqo chart values](/user/install/chart_values).

> Depending on the physical setup of your cluster, you need to properly configure some parameters required by Liqo during the install process to enable remote clusters to contact the above services. Below we present some common scenarios that Liqo can handle. Once you identify yours, you can go have to the *table* of each section to find the right values you should specify when installing Liqo using the chart.

Reviewer 2: ```suggestion
Depending on the physical setup of your cluster, you need to properly configure, at install time, how those services are exposed and can reached by remote clusters. Below we present some common scenarios that Liqo can handle. Once you identify yours, you can refer to the *table* of each section to find how to determine the right values you should specify.
```

The exposition parameters can be configured at installation time using the [Liqo Helm Chart](/user/install/chart_values) and updated after the installation by issuing an ``helm update`` after changing them in your values.yml. If you need more information about Helm and how charts can be configured, you can have a look at the [Helm official documentation](https://helm.sh/docs/). Pay attention that changing exposition parameters may affect and break active peerings. We suggest disabling all peerings before changing the way a cluster is exposed.

### Cloud to cloud

![](/images/scenarios/cloud-to-cloud.svg)

> Two managed clusters peered together through the Internet. It is possible to have a multi-cloud setup (AKS to AKS, GKE to GKE, and AKS to GKE). In this scenario, the services to exposes should be exposed using Public IPs/URLs.

Reviewer 2: ```suggestion
Two managed clusters peered together through the Internet. It is possible to have a multi-cloud setup (AKS to AKS, GKE to GKE, and AKS to GKE). In this scenario, the services should be exposed using Public IPs/URLs.
```

|  | Cluster A (Cloud) | Cluster B (Cloud) |
| --------- | -------- |  ---------       | |
| **Auth Server** |  LoadBalancer/ingress | LoadBalancer/Ingress |
| **API server** | Provided | Provided |

> | **Network gateway** | LoadBalancer/Node Port | LoadBalancer/Node Port |

Reviewer 2: Two comments here. 
1. The table seems broken (at least in the GitHub preview)
2. What does Provided mean?

> Considering the Network Gateway, at least one among Cluster A and Cluster B should have the **Network Gateway** IP accessible from the other one (e.g., Public IP).

Reviewer 2: ```suggestion
At least one among Cluster A and Cluster B should have the **Network Gateway** IP accessible from the other one (e.g., Public IP).
```

### On-premise to cloud

![](/images/scenarios/on-prem-to-cloud.svg)

On-premise cluster (K3s or K8s) exposed through the Internet peered with a Managed cluster (AKS or GKE).

|  | Cluster A (On-prem) | Cluster B (Cloud) |
| --------- | -------- |  ---------       |
| **Auth Server** |  LoadBalancer/Ingress | LoadBalancer/Ingress |
| **API server** | Ingress/Public IP | Provided |
| **Network gateway** | LoadBalancer/Node Port | LoadBalancer/Node Port |

Clusters API Server should be mutually accessible, and so should be for the Auth Service.

> Considering the Network Gateway, at least one among Cluster A and Cluster B should have the **Network Gateway** IP accessible from the other one (e.g., Public IP). If you configure the Auth service as Ingress, you should pay attention to disable TLS on the service or, more safely, to configure your Ingress Controller to support a TLS backend.

Reviewer 2: Would avoid the Network Gateway repetition. What does *more safely* mean?

#### On-premise behind NAT to cloud

![](/images/scenarios/on-prem-nat-to-cloud.svg)

When the On-premise cluster is exposed through a NAT, the configuration slightly changes:

|  | Cluster A (On-prem behind NAT) | Cluster B (Cloud) |
| --------- | -------- |  ---------       |
| **Auth Server** |  NodePort with port-forwarding | LoadBalancer/ingress |
| **API server** | Port-forwarding | Provided |
| **Network gateway** | NodePort with port-forwarding | LoadBalancer |

In this situation, the "cloud" cluster should have the Network Gateway exposed as a **LoadBalancer**. A couple of port-forwardings should also be configured for the Auth Server and K8s API Server to make them accessible from Cloud B.

### Clusters in the same LAN

![](/images/scenarios/on-prem-to-on-prem.svg)

Clusters (K3s or K8s) in the same LAN may rely on the mDNS-based Liqo discovery mechanism.

> The Liqo discovery mechanism based on mDNS will handle the discovery automatically. If you have your clusters in different L3 domains, you have to manually [create](/user/post-install/discovery#forging-the-foreigncluster) a *foreign_cluster* resource or rely on [DNS discovery](/user/post-install/discovery#manual-configuration).

Reviewer 2: Why is this written only here? Shouldn't it apply to all scenarios.
